documentation.md

```diff
@@ -0,0 +1,96 @@
+
+## Sequence Diagram
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant Caller as Caller
+    participant HandleTransaction as Handle Transaction
+    participant ExecuteRequest as Execute Request
+    participant EvaluateTypologyExpr as Evaluate Typology Expr
+    participant RedisCache as Redis Cache
+    participant Database as Database
+    participant CMS as CMS
+    participant CADP as CADP
+
+    Note over Caller, HandleTransaction: The process starts by handling a transaction.
+
+    Caller->>HandleTransaction: call(transaction, networkMap, ruleResult)
+    HandleTransaction->>+ExecuteRequest: call(transaction, typology, ruleResult, networkMap)
+
+    Note over ExecuteRequest: Execute Request processes the transaction for each typology.
+
+    loop channel.typologies
+        ExecuteRequest->>RedisCache: addOneGetAll(cacheKey, ruleResult)
+        ExecuteRequest->>RedisCache: check cached ruleResults
+
+        Note over RedisCache: Redis Cache stores interim rule results for faster processing.
+
+        opt no cached ruleResults
+            ExecuteRequest->>HandleTransaction: return CADPRequest
+        end
+        ExecuteRequest->>Database: getTypologyExpression(typology)
+
+        Note over Database: Database provides the typology expression needed for evaluation.
+
+        ExecuteRequest->>EvaluateTypologyExpr: call(ruleValues, ruleResults, typologyExpression)
+
+        Note over EvaluateTypologyExpr: Evaluate Typology Expr computes the typology result value.
+
+        opt typologyResultValue > expression.threshold
+            ExecuteRequest->>CMS: send Typology result
+        end
+
+        Note over CMS: CMS receives typology results that exceed the threshold.
+
+        ExecuteRequest->>CADP: send Typology result
+
+        Note over CADP: CADP receives all the typology results.
+
+        ExecuteRequest->>RedisCache: deleteKey(cacheKey)
+
+        Note over RedisCache: Redis Cache deletes the interim cache key after processing.
+
+        ExecuteRequest->>HandleTransaction: return CADPRequest
+    end
+
+    HandleTransaction-->>ExecuteRequest: CADPRequest returned
+    HandleTransaction->>HandleTransaction: add CADPRequest to CombinedResult.cadpRequests
+
+    Note over HandleTransaction: Handle Transaction accumulates the CADPRequests.
+
+    HandleTransaction->>Caller: return CombinedResult
+
+    Note over Caller: The final combined result is returned to the caller.
+
+
+```
+
+Here's a detailed explanation of the numbers in the sequence diagram:
+
+1. The process starts by handling a transaction. The caller initiates the request to process a transaction by calling the HandleTransaction function with the transaction, network map, and rule result as input parameters.
+2. The HandleTransaction function, in turn, calls the ExecuteRequest function for each typology in the transaction. It passes the transaction, typology, rule result, and network map as input parameters.
```

Comment:
"for each typology in the transaction" should read "for each typology in the network map associated with the rule".
Also, I assume the typologies for the rule are not apparent without some kind of processing. Because "work" is being done by the processor to determine the typologies, I would recommend that this step should be explicit in the diagram.

Reply:
Correct, there is a filter happening to get all typologies expecting this rule result.